unidist/core/backends/mpi/core/controller/common.py

```diff
@@ -175,39 +182,41 @@ def request_worker_data(data_id):
     """
     mpi_state = communication.MPIState.get_instance()
     local_store = LocalObjectStore.get_instance()
+    async_operations = AsyncOperations.get_instance()
 
-    owner_rank = local_store.get_data_owner(data_id)
+    for data_id in data_ids:
+        owner_rank = local_store.get_data_owner(data_id)
 
-    logger.debug("GET {} id from {} rank".format(data_id._id, owner_rank))
+        logger.debug("GET {} id from {} rank".format(data_id._id, owner_rank))
 
-    # Worker request
-    operation_type = common.Operation.GET
-    operation_data = {
-        "source": mpi_state.global_rank,
-        "id": data_id,
-        # set `is_blocking_op` to `True` to tell a worker
-        # to send the data directly to the requester
-        # without any delay
-        "is_blocking_op": True,
-    }
-    # We use a blocking send here because we have to wait for
-    # completion of the communication, which is necessary for the pipeline to continue.
-    communication.send_simple_operation(
-        mpi_state.comm,
-        operation_type,
-        operation_data,
-        owner_rank,
-    )
-
-    # Blocking get
-    complex_data = pull_data(mpi_state.comm, owner_rank)
-    if data_id != complex_data["id"]:
-        raise ValueError("Unexpected data_id!")
-    data = complex_data["data"]
-
-    # Caching the result, check the protocol correctness here
-    local_store.put(data_id, data)
-    return data
+        # Worker request
+        operation_type = common.Operation.GET
+        operation_data = {
+            "source": mpi_state.global_rank,
+            "id": data_id,
+            # set `is_blocking_op` to `True` to tell a worker
+            # to send the data directly to the requester
+            # without any delay
+            "is_blocking_op": True,
+        }
+        # We use a blocking send here because we have to wait for
+        # completion of the communication, which is necessary for the pipeline to continue.
+        h_list = communication.isend_simple_operation(
+            mpi_state.comm,
+            operation_type,
+            operation_data,
+            owner_rank,
+        )
+        async_operations.extend(h_list)
```

Comment:
We should probably wait on these async requests.

Reply:
No, because some data may be received before the sending operation is processed by another process.